File: app/app/src/main/java/project/picpay/test/home/view/HomeActivity.java
package project.picpay.test.home.view;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;
import android.widget.Toast;

import java.util.ArrayList;

import project.picpay.test.R;
import project.picpay.test.creditcard.view.ListCardsActivity;
import project.picpay.test.databinding.ActivityHomeBinding;
import project.picpay.test.home.model.UserModel;
import project.picpay.test.home.viewmodel.UserViewModel;
import project.picpay.test.transaction.TransactionActivity;

public class HomeActivity extends AppCompatActivity implements ListUserAdapter.OnActionListerner {

    private static final String TAG = HomeActivity.class.getSimpleName();
    private ActivityHomeBinding binding;
    private ListUserAdapter pullAdapter;
    private OnClickListerners listerners;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_home);

        configRecycleView();
        configListerners();
        configViewModel();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    public void onSelectUser(UserModel user) {
        Intent i = new Intent(HomeActivity.this, TransactionActivity.class);
        i.putExtra("user_model", user);
        startActivity(i);
    }

    private void configViewModel() {
        UserViewModel viewModel = ViewModelProviders.of(this).get(UserViewModel.class);

        viewModel.dataLoadStatus().observe(this, loadStatus -> {
            assert loadStatus != null;
            switch (loadStatus) {
                case LOADING:
                    //Toast.makeText(this, getString(R.string.s_message_loading_users), Toast.LENGTH_SHORT).show();
                    break;
                case LOADED:
                    Toast.makeText(this, getString(R.string.s_message_load_success_users), Toast.LENGTH_SHORT).show();
                    break;
                case FAILED:
                    showErrorList();
                    break;
            }
        });

        viewModel.getAvailableUsers().observe(HomeActivity.this, userModel -> {
            if (userModel == null) {
                return;
            }
            if (userModel.size() > 0) {
                pullAdapter.addItems(userModel);
                showRvImport();
            } else {
                showEmptyList();
            }
        });
    }

    private void configRecycleView() {
        pullAdapter = new ListUserAdapter(new ArrayList<>(), this);
        binding.rvMain.rvList.setLayoutManager(new LinearLayoutManager(this));
        binding.rvMain.rvList.setAdapter(pullAdapter);
    }

    private void configListerners() {
        listerners = new OnClickListerners(this);
        binding.setListerners(listerners);
    }

    private void showRvImport() {
        binding.loadMain.pbLoad.setVisibility(View.GONE);
        binding.tvMain.llEmptyValues.setVisibility(View.GONE);
        binding.rvMain.rvList.setVisibility(View.VISIBLE);
    }

    private void showEmptyList() {
        binding.loadMain.pbLoad.setVisibility(View.GONE);
        binding.rvMain.rvList.setVisibility(View.GONE);
        binding.tvMain.tvNoValues.setText(getString(R.string.s_message_without_users));
        binding.tvMain.llEmptyValues.setVisibility(View.VISIBLE);
    }

    private void showErrorList() {
        binding.loadMain.pbLoad.setVisibility(View.GONE);
        binding.rvMain.rvList.setVisibility(View.GONE);
        binding.tvMain.tvNoValues.setText(getString(R.string.s_message_error_users));
        binding.tvMain.llEmptyValues.setVisibility(View.VISIBLE);
    }

    public class OnClickListerners {

        Context context;

        OnClickListerners(Context context) {
            this.context = context;
        }

        public void newCreditCard(View view) {

        }

        public void listCreditCards(View view) {
            startActivity(new Intent(HomeActivity.this, ListCardsActivity.class));
        }

    }

    /*private void showDialogTransaction(UserModel user) {
        Dialog dialog = new Dialog(this);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
        DialogTransactionBinding binding = DataBindingUtil.inflate(LayoutInflater.from(this), R.layout.dialog_transaction, null, false);
        binding.setListerners(listerners);
        binding.setPost(user);
        dialog.setContentView(binding.getRoot());
        dialog.show();
    }*/

}

File: app/src/main/java/test/edney/picpay/view/home/HomeActivity.kt
package test.edney.picpay.view.home

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View.*
import android.view.animation.AnimationUtils
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.gson.Gson
import test.edney.picpay.R
import test.edney.picpay.view.ReceiptDialog
import test.edney.picpay.databinding.ActivityHomeBinding
import test.edney.picpay.model.UserModel
import test.edney.picpay.view.payment.PaymentActivity
import android.view.inputmethod.InputMethodManager
import android.widget.RelativeLayout
import android.animation.LayoutTransition
import test.edney.picpay.util.ExtrasName
import test.edney.picpay.view.card.CardActivity
import test.edney.picpay.viewmodel.HomeVM

class HomeActivity : AppCompatActivity() {

      private lateinit var binding: ActivityHomeBinding
      private lateinit var viewmodel: HomeVM
      private lateinit var mUserAdapter: UserAdapter
      private val receiptDelay = 400L
      private val mGson = Gson()

      override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)

            binding = DataBindingUtil.setContentView(this, R.layout.activity_home)
            viewmodel()
            ui()

            showReceipt()
            configureSearch()
            configureListOfUser()
      }

      private fun viewmodel() {
            viewmodel = ViewModelProviders.of(this).get(HomeVM::class.java)

            viewmodel.userResponse.observe(this, Observer {
                  if (it != null) {
                        mUserAdapter.postUsers(it)
                        stopLoading()
                  } else {
                        Toast.makeText(this@HomeActivity, getString(R.string.s_main_message_no_user), Toast.LENGTH_LONG)
                              .show()
                  }
            })
      }

      private fun ui(){
            binding.ui = object : HomeUI{
                  override fun actionSearch() {
                       if(!binding.etSearch.isFocused)
                             binding.etSearch.requestFocus()
                  }
                  override fun actionCloseSearch() {
                        binding.etSearch.setText("")
                  }
            }
      }

      private fun configureSearch(){
            binding.llSearch.layoutTransition = LayoutTransition()

            binding.etSearch.addTextChangedListener(object : TextWatcher {
                  override fun afterTextChanged(s: Editable?) {
                        val text = s.toString()

                        mUserAdapter.getFilter().filter(text)
                  }
                  override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
                  override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            })
            binding.etSearch.setOnFocusChangeListener { _, hasFocus ->
                  val layout = RelativeLayout.LayoutParams(
                        if(hasFocus) RelativeLayout.LayoutParams.MATCH_PARENT
                        else RelativeLayout.LayoutParams.WRAP_CONTENT,
                        RelativeLayout.LayoutParams.WRAP_CONTENT
                  )

                  binding.llSearch.layoutParams = layout
                  binding.llSearch.requestLayout()

                  binding.rlSearch.setBackgroundResource(
                        if(hasFocus) R.drawable.edittext_focused else R.drawable.edittext
                  )

                  binding.ivSearch.setColorFilter(ContextCompat.getColor(this,
                        if(hasFocus) R.color.colorWhite else R.color.colorSearchText)
                  )

                  binding.ivSearchClose.visibility = if(hasFocus) VISIBLE else GONE
            }
            binding.etSearch.setOnKeyListener { _, keyCode, event ->
                  var stop = false

                  if (keyCode == KeyEvent.KEYCODE_BACK
                        && event?.action == KeyEvent.ACTION_UP
                  ) {
                        removeSearchFocus()
                        stop = true
                  }
                  stop
            }
      }

      private fun showReceipt() {
            val args = intent.getStringExtra(ExtrasName.transaction)

            if (args != null) {
                  val dialog = ReceiptDialog()
                  val argsToDialog = Bundle()
                  val cardNumber = intent.getStringExtra(ExtrasName.card_number)

                  if(cardNumber != null)
                        argsToDialog.putString(ExtrasName.card_number, cardNumber)
                  argsToDialog.putString(ExtrasName.transaction, args)
                  dialog.arguments = argsToDialog
                  Handler().postDelayed({ dialog.show(supportFragmentManager, "receipt_dialog") }, receiptDelay)
            }
      }

      private fun configureListOfUser() {
            val animation = AnimationUtils.loadLayoutAnimation(this@HomeActivity, R.anim.layout_anim)

            mUserAdapter = UserAdapter(object : UserAdapterListener {
                  override fun onItemClick(user: UserModel) {
                        val hasCard = viewmodel.hasCard()
                        val intent: Intent

                        if(hasCard){
                              intent = Intent(this@HomeActivity, PaymentActivity::class.java).apply {
                                    putExtra(ExtrasName.has_card, hasCard)
                                    putExtra(ExtrasName.user, mGson.toJson(user))
                              }
                        }
                        else{
                              intent = Intent(this@HomeActivity, CardActivity::class.java).apply {
                                    putExtra(ExtrasName.has_card, hasCard)
                                    putExtra(ExtrasName.user, mGson.toJson(user))
                              }
                        }
                        startActivity(intent)
                  }
            })
            binding.rvUser.setHasFixedSize(true)
            binding.rvUser.layoutAnimation = animation
            binding.rvUser.layoutManager = LinearLayoutManager(this)
            binding.rvUser.adapter = mUserAdapter
            viewmodel.requestUsers()
      }

      private fun stopLoading() {
            binding.progress.visibility = GONE
            binding.rvUser.visibility = VISIBLE
      }

      private fun removeSearchFocus(){
            binding.etSearch.clearFocus()
            if (currentFocus != null) {
                  val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                  imm.hideSoftInputFromWindow(currentFocus?.windowToken, 0)
            }
      }
}